Let env_method accept indices for SubprocVecEnv and DummyVecEnv

kantneel opened the pull request:
There exist situations in which one wishes to call a method in only a subset of environments inside of SubprocVecEnv and/or DummyVecEnv. The `set_attr` methods for both of these classes provide an example of how to do this with an `indices` argument. These changes replicate that logic for the `env_method` method to achieve the desired result.

araffin commented:
Hello,
Thank for this PR. I recommend you reading the `CONTRIBUTING.md` guide ;) (for this time, you can skip creating the issue, but would be appreciated next time)

>The set_attr methods for both of these classes provide an example of how to do this with an indices argument.

I would do that for the `get_attr` method too then. Btw, I would refactor a bit, creating a `_get_indices()` class method (maybe in the base class) to avoid duplicating code.

kantneel commented:
Okay, I will refactor the code as you suggested. The PR appears to have broken a couple of tests so I will address that as well. Thanks!

AdamGleave commented:
:+1: to adding a `_get_indices` helper method.

I also notice we don't have `get_attr` or `set_attr` in `BaseVecEnv`. It seems worth adding these as abstract methods. It's possible we'll want to add some more `VecEnv`'s in the future, e.g. the shared memory one from Baselines, but this doesn't seem like a hard feature to add to them.

kantneel commented:
So I've refactored everything in accordance with the conversation thus far. However, I'm not completely satisfied with the call signature of `env_method`. It's annoying that `indices` cannot be used as a keyword argument since we also have `**method_kwargs` and its use as a positional argument is confusing with `*method_args`. A demonstration of this is present in the test case that I modified in commit 6e6d2bb. Do you guys have any suggestions for this?

AdamGleave commented on stable_baselines/common/vec_env/dummy_vec_env.py:
```diff
@@ -47,7 +47,8 @@ def reset(self):
 
     def close(self):
-        return
+        for env in self.envs:
```
Wow! I'm surprised this was a no-op before. :+1:

AdamGleave commented on stable_baselines/common/vec_env/dummy_vec_env.py:
```diff
@@ -68,37 +69,19 @@ def _save_obs(self, env_idx, obs):
-        :return: (list) List of items retured by the environment's method call
-        """
-        return [getattr(env_i, method_name)(*method_args, **method_kwargs) for env_i in self.envs]
+    def env_method(self, method_name, indices=None, *method_args, **method_kwargs):
```
Needs documentation. I'd be OK with a one-sentence description pointing to the VecEnv base class (although @erniejunior may be more stringent), but it shouldn't be empty.

AdamGleave commented on stable_baselines/common/vec_env/dummy_vec_env.py:
```diff
@@ -69,37 +69,19 @@ def _save_obs(self, env_idx, obs):
+        target_envs = self._get_target_envs(indices)
+        return [getattr(env_i, method_name)(*method_args, **method_kwargs) for env_i in target_envs]
+
+    def get_attr(self, attr_name, indices=None):
```
Needs doc.

AdamGleave commented on stable_baselines/common/vec_env/dummy_vec_env.py:
```diff
@@ -69,37 +69,19 @@ def _save_obs(self, env_idx, obs):
+        target_envs = self._get_target_envs(indices)
+        return [getattr(env_i, attr_name) for env_i in target_envs]
 
     def set_attr(self, attr_name, value, indices=None):
```
Needs doc.

AdamGleave commented on stable_baselines/common/vec_env/subproc_vec_env.py:
```diff
@@ -141,52 +141,36 @@ def get_images(self):
-        """
-
-        for remote in self.remotes:
+    def env_method(self, method_name, indices=None, *method_args, **method_kwargs):
```
Needs doc.

AdamGleave commented on stable_baselines/common/vec_env/subproc_vec_env.py:
```diff
@@ -141,52 +141,36 @@ def get_images(self):
+    def get_attr(self, attr_name, indices=None):
         """
-        Provides a mechanism for getting class attribues from vectorized environments
+        Provides a mechanism for getting class attributes from vectorized environments
```
Doc should reference base class.

AdamGleave commented on stable_baselines/common/vec_env/subproc_vec_env.py:
```diff
@@ -141,52 +141,36 @@ def get_images(self):
-        return [remote.recv() for remote in self.remotes]
+        return [remote.recv() for remote in target_remotes]
 
     def set_attr(self, attr_name, value, indices=None):
```
Doc should reference base class.

AdamGleave commented on stable_baselines/common/vec_env/base_vec_env.py:
```diff
@@ -83,6 +83,42 @@ def close(self):
+        pass
+
+    @abstractmethod
+    def env_method(self, method_name, indices=None, *method_args, **method_kwargs):
```
How about changing signature to:
`def env_method(self, method_name, *method_args,  indices=None, **method_kwargs):`
This would make `indices` only accessible as a keyword-argument, but this seems maybe desirable, and certainly better than breaking the API for existing users of `env_method`.

araffin commented:
It starts to look good. Please do not forget to:
- update the changelog (and put your name at the bottom)
- add tests to ensure that the new features work properly

erniejunior commented:
> However, I'm not completely satisfied with the call signature of `env_method`. It's annoying that `indices` cannot be used as a keyword argument since we also have `**method_kwargs` and its use as a positional argument is confusing with `*method_args`

I have two ideas here. Feel free to vigorously reject them because they introduce additional complexity.
1. Make `VecEnv` implement `__getitem__`. This allows to first select a subset of environments (examples: `vec_env[2]`, `vec_env[2:5]`, `vec_env[[2, 5, 7]]`) and then call the methods on those subsets with the same API that we use now (e.g. `vec_env[2:5].env_method(...)`).
2. Extends the API to allow to transparently call the environment method by using some `__getattribute__` magic. For example this class allows you to call arbitrary methods with arbitrary parameters:
```python
class B:
    def __getattribute__(self, name):
        def m(*args, **kwargs):
            print(name, "called with", args, kwargs)
        return m
b = B()
b.my_fancy_method(1, 2, 3, b=25)  # my_fancy_method called with (1, 2, 3) {'b': 25}
```

Here a simple mock-up of these two ideas for the DummyVecEnv case:
```python
class EnvSelection:
    def __init__(self, envs):
        self.envs=envs

    def __getattribute__(self, name):
        def call_methods(*args, **kwargs):
            return [getattr(env, name)(*args, **kwargs) for env in self.envs]
        return call_methods

class VecEnv:
    def __init__(self):
        self.envs = ...

    def __getitem__(self, item):
        return EnvSelection(self.envs[item])
```
This allows for calls like `my_vec_env[3:5].the_env_method(25, 26, b=7)`.

araffin commented:
@erniejunior sometimes magic can be helpful, however, here, I feel this add more complexity than needed and also hides too much what is happening.

erniejunior commented:
Thats probably true here :-)

araffin commented:
Almost there, I think the code is ready, but still missing:
- updating the changelog
- updating the tests to ensure that the new features work

kantneel commented:
@araffin thanks, I will work on #267 instead then.

AdamGleave commented:
Agreed we should add tests, @kantneel let me know when you've done that and I'll take another look at the code.

araffin commented:
@AdamGleave, it seems I cannot push to this branch :/ (ERROR: Permission to HumanCompatibleAI/baselines.git denied to araffin.)
So I've pushed the updates to `env_method` branch in stable-baselines repo. I let you retrieve the changes ;)

AdamGleave commented:
Thanks for the test cases and doc fixes @araffin. I've merged them now. (Not sure why you didn't have permission to push to this branch.)

I've also made some fixes to the docs, and a minor API change: `set_attr` now returns `None` rather than a list of `None`'s (since `setattr` should always return `None`).

Waiting for the tests to finish, will merge once complete.
